tests/scripts/all.sh

```diff
@@ -769,6 +782,95 @@ component_build_default_make_gcc_and_cxx () {
+    if_build_succeeded not test -f library/libmbedcrypto.a
+    msg "test: libmbedcrypto symbols are from crypto files (USE_CRYPTO_SUBMODULE, cmake)"
+    if_build_succeeded objdump -g crypto/library/libmbedcrypto.a | grep -E 'crypto/library$' > /dev/null
+    msg "test: libmbedcrypto uses top-level config (USE_CRYPTO_SUBMODULE, cmake)"
+    if_build_succeeded objdump -g crypto/library/libmbedcrypto.a | grep 'md4.c' > /dev/null
+    msg "test: main suites (USE_CRYPTO_SUBMODULE, cmake)"
+    make test
+    msg "test: ssl-opt.sh (USE_CRYPTO_SUBMODULE, cmake)"
+    if_build_succeeded tests/ssl-opt.sh
+}
+
+component_test_submodule_make () {
+    # USE_CRYPTO_SUBMODULE: check that the build works with make
+    msg "build: make, full config + USE_CRYPTO_SUBMODULE, gcc+debug"
+    scripts/config.pl full # enables md4 and submodule doesn't enable md4
+    scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE # too slow for tests
+    make CC=gcc CFLAGS='-g' USE_CRYPTO_SUBMODULE=1
+    msg "test: top-level libmbedcrypto wasn't built (USE_CRYPTO_SUBMODULE, make)"
+    if_build_succeeded not test -f library/libmbedcrypto.a
+    msg "test: libmbedcrypto symbols are from crypto files (USE_CRYPTO_SUBMODULE, make)"
+    if_build_succeeded objdump -g crypto/library/libmbedcrypto.a | grep -E 'crypto/library$' > /dev/null
+    msg "test: libmbedcrypto uses top-level config (USE_CRYPTO_SUBMODULE, make)"
+    if_build_succeeded objdump -g crypto/library/libmbedcrypto.a | grep 'md4.c' > /dev/null
+    msg "test: main suites (USE_CRYPTO_SUBMODULE, make)"
+    make CC=gcc USE_CRYPTO_SUBMODULE=1 test
+    msg "test: ssl-opt.sh (USE_CRYPTO_SUBMODULE, make)"
+    if_build_succeeded tests/ssl-opt.sh
+}
+
+component_test_no_submodule_make () {
+    # Don't USE_CRYPTO_SUBMODULE: check that the submodule is not used with make
+    msg "build: make, full config - USE_CRYPTO_SUBMODULE, gcc+debug"
+    scripts/config.pl full
+    make CC=gcc CFLAGS='-g'
+    msg "test: submodule libmbedcrypto wasn't built (USE_CRYPTO_SUBMODULE, make)"
+    if_build_succeeded not test -f crypto/library/libmbedcrypto.a
+    msg "test: libmbedcrypto symbols are from library files (USE_CRYPTO_SUBMODULE, make)"
+    if_build_succeeded objdump -g library/libmbedcrypto.a | grep -E 'library$' | not grep 'crypto' > /dev/null
+}
+
+component_test_no_submodule_cmake () {
+    # Don't USE_CRYPTO_SUBMODULE: check that the submodule is not used with CMake
+    msg "build: cmake, full config - USE_CRYPTO_SUBMODULE, gcc+debug"
+    scripts/config.pl full
+    CC=gcc cmake -D CMAKE_BUILD_TYPE=Debug .
+    make
+    msg "test: submodule libmbedcrypto wasn't built (USE_CRYPTO_SUBMODULE, cmake)"
+    if_build_succeeded not test -f crypto/library/libmbedcrypto.a
+    msg "test: libmbedcrypto symbols are from library files (USE_CRYPTO_SUBMODULE, cmake)"
+    if_build_succeeded objdump -g library/libmbedcrypto.a | grep -E 'library$' | not grep 'crypto' > /dev/null
+}
+
+component_test_submodule_full_cmake_asan() {
```

Comment:
I think the name should have `use_psa_crypto` instead of `submodule`, as the point is really to test that option, and the submodule is just a means to that end.

Reply:
Using `submodule` in the name has the advantage that you can run `all.sh '*submodule*'` to run all submodule-related tests.